File: src/LagrangianDAO.cpp
#include <symengine/add.h>
#include <symengine/mul.h>
#include <symengine/number.h>
#include <symengine/matrices/trace.h>
#include <symengine/symengine_exception.h>

#include "SymResponse/LagrangianDAO.hpp"

namespace SymResponse
{
    LagrangianDAO::LagrangianDAO(
        const SymEngine::RCP<const Tinned::Perturbation>& a,
        const SymEngine::RCP<const Tinned::OneElecDensity>& D,
        const SymEngine::RCP<const Tinned::OneElecOperator>& S,
        const SymEngine::vec_basic& H,
        const SymEngine::RCP<const Tinned::TwoElecOperator>& G,
        const SymEngine::RCP<const Tinned::ExchCorrEnergy>& Exc,
        const SymEngine::RCP<const Tinned::ExchCorrPotential>& Fxc,
        const SymEngine::RCP<const Tinned::NonElecFunction>& hnuc,
        const bool sym_elimination
    ) : sym_elimination_(sym_elimination), a_(a), D_(D), S_(S)
    {
        //FIXME: to implement for both cases: with/without intensive perturbations
        if (sym_elimination_) throw SymEngine::SymEngineException(
            "LagrangianDAO has not implemented for almost symmetric form of elimination!"
        );

        SymEngine::vec_basic E_terms;
        SymEngine::vec_basic F_terms;
        if (!H.empty()) {
            //E_terms.push_back(SymEngine::trace(
            //    SymEngine::matrix_mul({SymEngine::matrix_add(H), D})
            //));
            for (const auto& op: H)
                E_terms.push_back(SymEngine::trace(SymEngine::matrix_mul({op, D})));
            F_terms = H;
        }
        if (!G.is_null()) {
            E_terms.push_back(Tinned::make_2el_energy(G));
            F_terms.push_back(G);
        }
        if (!Exc.is_null()) E_terms.push_back(Exc);
        if (!Fxc.is_null()) F_terms.push_back(Fxc);
        if (!hnuc.is_null()) E_terms.push_back(hnuc);

        if (E_terms.empty()) throw SymEngine::SymEngineException(
            "LagrangianDAO gets nothing for generalized energy!"
        );
        if (F_terms.empty()) throw SymEngine::SymEngineException(
            "LagrangianDAO gets nothing for generalized Fock operator!"
        );

        E_ = SymEngine::add(E_terms);
        F_ = SymEngine::matrix_add(F_terms);

        // |i\frac{\partial `D_`}{\partial t}>
        auto Dt = Tinned::make_dt_operator(D);
        auto D_a = D->diff(a);
        // `Tinned::differentiate` can remove zero quantities after differentiation
        auto F_a = Tinned::differentiate(F_, Tinned::PertMultichain({a}));

        // Terms of differentiated quasi-energy derivative Lagrangian with
        // respect to the perturbation `a`
        SymEngine::vec_basic La_terms;

        auto one_half = SymEngine::div(SymEngine::one, SymEngine::two);
        auto minus_one_half = SymEngine::div(SymEngine::minus_one, SymEngine::two);
        // Null overlap operator means it is an identity operator
        if (S.is_null()) {
            // = D^{a}D-DD^{a}
            lambda_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({D_a, D}),
                SymEngine::matrix_mul({SymEngine::minus_one, D, D_a})
            });
            // Y = FD-DF-i\frac{\partial D}{\partial t}
            Y_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({F_, D}),
                SymEngine::matrix_mul({SymEngine::minus_one, D, F_}),
                SymEngine::matrix_mul({SymEngine::minus_one, Dt})
            });
            // = F^{a}D+DF^{a}-F^{a}
            zeta_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({F_a, D}),
                SymEngine::matrix_mul({D, F_a}),
                SymEngine::matrix_mul({SymEngine::minus_one, F_a})
            });
            // Z = D*D-D
            Z_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({D, D}),
                SymEngine::matrix_mul({SymEngine::minus_one, D})
            });
        }
        else {
            // |i\frac{\partial `S`}{\partial t}>
            auto St = Tinned::make_dt_operator(S);
            // Equation (95), J. Chem. Phys. 129, 214108 (2008)
            W_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({D, F_, D}),
                SymEngine::matrix_mul({one_half, Dt, S, D}),
                SymEngine::matrix_mul({minus_one_half, D, S, Dt})
            });
            // Pulay term
            auto S_a = S->diff(a);
            if (!Tinned::is_zero_quantity(S_a)) La_terms.push_back(
                SymEngine::trace(SymEngine::matrix_mul({S_a, W_}))
            );
            // Equation (220), J. Chem. Phys. 129, 214108 (2008)
            lambda_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({D_a, S, D}),
                SymEngine::matrix_mul({SymEngine::minus_one, D, S, D_a})
            });
            // Equation (229), J. Chem. Phys. 129, 214108 (2008)
            Y_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({F_, D, S}),
                SymEngine::matrix_mul({SymEngine::minus_one, S, D, F_}),
                SymEngine::matrix_mul({SymEngine::minus_one, S, Dt, S}),
                SymEngine::matrix_mul({minus_one_half, St, D, S}),
                SymEngine::matrix_mul({minus_one_half, S, D, St})
            });
            // Equation (224), J. Chem. Phys. 129, 214108 (2008)
            zeta_ = Tinned::is_zero_quantity(S_a)
                  ? SymEngine::matrix_add({
                        SymEngine::matrix_mul({F_a, D, S}),
                        SymEngine::matrix_mul({S, D, F_a}),
                        SymEngine::matrix_mul({SymEngine::minus_one, F_a})
                    })
                  : SymEngine::matrix_add({
                        SymEngine::matrix_mul({F_a, D, S}),
                        SymEngine::matrix_mul({SymEngine::minus_one, F_, D, S_a}),
                        SymEngine::matrix_mul({one_half, St, D, S_a}),
                        SymEngine::matrix_mul({S, Dt, S_a}),
                        SymEngine::matrix_mul({S, D, F_a}),
                        SymEngine::matrix_mul({SymEngine::minus_one, S_a, D, F_}),
                        SymEngine::matrix_mul({minus_one_half, S_a, D, St}),
                        SymEngine::matrix_mul({SymEngine::minus_one, S_a, Dt, S}),
                        SymEngine::matrix_mul({SymEngine::minus_one, F_a})
                    });
            // Equation (230), J. Chem. Phys. 129, 214108 (2008)
            Z_ = SymEngine::matrix_add({
                SymEngine::matrix_mul({D, S, D}),
                SymEngine::matrix_mul({SymEngine::minus_one, D})
            });
        }
        // Make an "artificial" Lagrangian multiplier for elimination
        tdscf_multiplier_ = Tinned::make_perturbed_parameter(
            std::string("tdscf-multiplier")
        );
        La_terms.push_back(
            SymEngine::trace(SymEngine::matrix_mul({tdscf_multiplier_, Y_}))
        );
        // Make an "artificial" Lagrangian multiplier for elimination
        idempotency_multiplier_ = Tinned::make_perturbed_parameter(
            std::string("idempotency-multiplier")
        );
        La_terms.push_back(
            SymEngine::trace(SymEngine::matrix_mul({idempotency_multiplier_, Z_}))
        );
        // Construct the time-averaged quasi-energy derivative Lagrangian
        La_ = SymEngine::add({
            // The first term in Equation (98), J. Chem. Phys. 129, 214108 (2008)
            Tinned::remove_if(
                Tinned::differentiate(E_, Tinned::PertMultichain({a_})),
                SymEngine::set_basic({D_->diff(a_)})
            ),
            SymEngine::mul(SymEngine::minus_one, SymEngine::add(La_terms))
        });
    }

    bool LagrangianDAO::validate_perturbation_frequencies(
        const Tinned::PertMultichain& exten_perturbations,
        const Tinned::PertMultichain& inten_perturbations,
        const SymEngine::RCP<const SymEngine::Number>& threshold
    ) const noexcept
    {
        // Here we need to inlcude the frequency of the perturbation `a_`,
        // which can either be extensive or intensive
        SymEngine::RCP<const SymEngine::Basic> sum_freq = a_->get_frequency();
        for (const auto& p: exten_perturbations)
            sum_freq = SymEngine::add(sum_freq, p->get_frequency());
        for (const auto& p: inten_perturbations)
            sum_freq = SymEngine::add(sum_freq, p->get_frequency());
        if (SymEngine::is_a_Number(*sum_freq)) {
            return Tinned::is_zero_number(
                SymEngine::rcp_dynamic_cast<const SymEngine::Number>(sum_freq),
                threshold
            );
        }
        else {
            // For nonnumerical frequencies, we simply return `true` and users
            // are responsible for the validation
            return true;
        }
    }

    SymEngine::RCP<const SymEngine::Basic> LagrangianDAO::get_lagrangian() const noexcept
    {
        return La_;
    }

    SymEngine::RCP<const SymEngine::Basic> LagrangianDAO::eliminate_wavefunction_parameter(
        const SymEngine::RCP<const SymEngine::Basic>& L,
        const Tinned::PertMultichain& exten_perturbations,
        const unsigned int min_wfn_order
    )
    {
        return Tinned::eliminate(L, D_, exten_perturbations, min_wfn_order);
    }

    SymEngine::RCP<const SymEngine::Basic> LagrangianDAO::eliminate_lagrangian_multipliers(
        const SymEngine::RCP<const SymEngine::Basic>& L,
        const Tinned::PertMultichain& exten_perturbations,
        const unsigned int min_multiplier_order
    )
    {
        auto result = Tinned::eliminate(
            L, tdscf_multiplier_, exten_perturbations, min_multiplier_order
        );
        result = Tinned::eliminate(
            result, idempotency_multiplier_, exten_perturbations, min_multiplier_order
        );
        // Replace "artificial" multipliers with real differentiated ones
        return Tinned::replace_all<Tinned::PerturbedParameter>(
            result,
            Tinned::TinnedBasicMap<Tinned::PerturbedParameter>({
                {tdscf_multiplier_, lambda_}, {idempotency_multiplier_, zeta_}
            })
        );
    }

    //TODO: double residue problem happens for Hessian of excited states
}

File: src/lagrangian_dao.rs
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

use num_rational::Rational64;

use typetag;

use tinned::{
    Add, Expr, LagMultiplier, MatrixAdd, MatrixMul, Number, NumberTolerance, Perturbation,
    TemporumOperator, TinnedError, Trace, TwoElecEnergy, TwoElecOperator, WfnParameter,
    anticommutator, commutator, differentiate_expr, downcast_from_arc, expression_error,
    generic_error, is_expr_type, is_zero_expr, s_anticommutator, s_commutator, subtract_exprs,
    sum_pert_frequencies,
};

use crate::lagrangian::Lagrangian;
use crate::lagrangian_internal::sealed::LagrangianInternal;

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct LagrangianDao {
    perturbation_a: Option<Arc<Perturbation>>,
    density_matrix: Arc<dyn Expr>,
    overlap_matrix: Option<Arc<dyn Expr>>,
    //generalized_energy: Arc<dyn Expr>,
    //fock_matrix: Arc<dyn Expr>,
    //ewd_matrix: Arc<dyn Expr>,
    lambda: Arc<dyn Expr>,
    tdscf_multiplier: Arc<dyn Expr>,
    tdscf_equation: Arc<dyn Expr>,
    zeta: Arc<dyn Expr>,
    idemp_multiplier: Arc<dyn Expr>,
    idempotency: Arc<dyn Expr>,
    lagrangian_dao: Arc<dyn Expr>,
}

impl LagrangianDao {
    pub fn new(
        perturbation_a: Arc<Perturbation>,
        density_matrix: Arc<dyn Expr>,
        overlap_matrix: Option<Arc<dyn Expr>>,
        one_elec_operators: &[Arc<dyn Expr>],
        two_elec_operator: Option<Arc<dyn Expr>>,
        xc_energy: Option<Arc<dyn Expr>>,
        xc_potential: Option<Arc<dyn Expr>>,
        h_nuc: Option<Arc<dyn Expr>>,
        num_tol: Option<NumberTolerance>,
    ) -> Result<Self, TinnedError> {
        let (generalized_energy, fock_matrix) = Self::build_energy_and_fock(
            density_matrix.clone(),
            one_elec_operators,
            two_elec_operator,
            xc_energy,
            xc_potential,
            h_nuc,
        )?;

        // |i\frac{\partial `D_`}{\partial t}>
        let density_t = TemporumOperator::builder(density_matrix.clone())
            .is_forward(true)
            .build()?;
        let density_a = density_matrix.differentiate(&perturbation_a)?;
        let fock_a = fock_matrix.differentiate(&perturbation_a)?;

        // Constraint and Lagrangian terms
        let mut lag_terms = Vec::with_capacity(3);

        let (lambda, tdscf_equation, zeta, idempotency) = if let Some(overlap) = &overlap_matrix {
            // |i\frac{\partial `S`}{\partial t}>
            let overlap_t = TemporumOperator::builder(overlap.clone())
                .is_forward(true)
                .build()?;
            let minus_one = Number::minus_one();
            let one_half = Number::one_half();
            let negative_one_half = Number::from_rational(Rational64::new(-1, 2));

            // Equation (220), J. Chem. Phys. 129, 214108 (2008)
            let lambda = s_commutator(density_a.clone(), density_matrix.clone(), overlap.clone())?;

            // Equation (229), J. Chem. Phys. 129, 214108 (2008)
            let tdscf_equation = MatrixAdd::new(vec![
                MatrixMul::new(vec![
                    fock_matrix.clone(),
                    density_matrix.clone(),
                    overlap.clone(),
                ])?,
                MatrixMul::new(vec![
                    minus_one.clone(),
                    overlap.clone(),
                    density_matrix.clone(),
                    fock_matrix.clone(),
                ])?,
                MatrixMul::new(vec![
                    minus_one.clone(),
                    overlap.clone(),
                    density_t.clone(),
                    overlap.clone(),
                ])?,
                MatrixMul::new(vec![
                    negative_one_half.clone(),
                    overlap_t.clone(),
                    density_matrix.clone(),
                    overlap.clone(),
                ])?,
                MatrixMul::new(vec![
                    negative_one_half.clone(),
                    overlap.clone(),
                    density_matrix.clone(),
                    overlap_t.clone(),
                ])?,
            ])?;

            // Equation (224), J. Chem. Phys. 129, 214108 (2008)
            let overlap_a = overlap.differentiate(&perturbation_a)?;
            let zeta = if is_zero_expr(&overlap_a, num_tol) {
                MatrixAdd::new(vec![
                    MatrixMul::new(vec![
                        fock_a.clone(),
                        density_matrix.clone(),
                        overlap.clone(),
                    ])?,
                    MatrixMul::new(vec![
                        overlap.clone(),
                        density_matrix.clone(),
                        fock_a.clone(),
                    ])?,
                    MatrixMul::new(vec![minus_one.clone(), fock_a])?,
                ])?
            } else {
                // Equation (95), J. Chem. Phys. 129, 214108 (2008)
                let ewd_matrix = MatrixAdd::new(vec![
                    MatrixMul::new(vec![
                        density_matrix.clone(),
                        fock_matrix.clone(),
                        density_matrix.clone(),
                    ])?,
                    MatrixMul::new(vec![
                        one_half.clone(),
                        density_t.clone(),
                        overlap.clone(),
                        density_matrix.clone(),
                    ])?,
                    MatrixMul::new(vec![
                        negative_one_half.clone(),
                        density_matrix.clone(),
                        overlap.clone(),
                        density_t.clone(),
                    ])?,
                ])?;
                // Pulay term
                lag_terms.push(Trace::new(MatrixMul::new(vec![
                    overlap_a.clone(),
                    ewd_matrix,
                ])?)?);

                MatrixAdd::new(vec![
                    MatrixMul::new(vec![
                        fock_a.clone(),
                        density_matrix.clone(),
                        overlap.clone(),
                    ])?,
                    MatrixMul::new(vec![
                        minus_one.clone(),
                        fock_matrix.clone(),
                        density_matrix.clone(),
                        overlap_a.clone(),
                    ])?,
                    MatrixMul::new(vec![
                        one_half.clone(),
                        overlap_t.clone(),
                        density_matrix.clone(),
                        overlap_a.clone(),
                    ])?,
                    MatrixMul::new(vec![overlap.clone(), density_t.clone(), overlap_a.clone()])?,
                    MatrixMul::new(vec![
                        overlap.clone(),
                        density_matrix.clone(),
                        fock_a.clone(),
                    ])?,
                    MatrixMul::new(vec![
                        minus_one.clone(),
                        overlap_a.clone(),
                        density_matrix.clone(),
                        fock_matrix,
                    ])?,
                    MatrixMul::new(vec![
                        negative_one_half,
                        overlap_a.clone(),
                        density_matrix.clone(),
                        overlap_t,
                    ])?,
                    MatrixMul::new(vec![
                        minus_one.clone(),
                        overlap_a,
                        density_t,
                        overlap.clone(),
                    ])?,
                    MatrixMul::new(vec![minus_one, fock_a])?,
                ])?
            };

            // First term of Equation (230), J. Chem. Phys. 129, 214108 (2008)
            let idempotency = MatrixMul::new(vec![
                density_matrix.clone(),
                overlap.clone(),
                density_matrix.clone(),
            ])?;

            (lambda, tdscf_equation, zeta, idempotency)
        } else {
            // = D^{a}D-DD^{a}
            let lambda = commutator(density_a.clone(), density_matrix.clone())?;
            // Y = FD-DF-i\frac{\partial D}{\partial t}
            let tdscf_equation = subtract_exprs(
                commutator(fock_matrix.clone(), density_matrix.clone())?,
                density_t,
            )?;
            // = F^{a}D+DF^{a}-F^{a}
            let zeta = subtract_exprs(
                anticommutator(fock_a.clone(), density_matrix.clone())?,
                fock_a,
            )?;
            // = D*D
            let idempotency = MatrixMul::new(vec![density_matrix.clone(), density_matrix.clone()])?;

            (lambda, tdscf_equation, zeta, idempotency)
        };

        // Make "artificial" Lagrangian multipliers for elimination
        let tdscf_multiplier = LagMultiplier::builder("tdscf-multiplier").build()?;
        let idemp_multiplier = LagMultiplier::builder("idemp_multiplier").build()?;

        lag_terms.push(Trace::new(MatrixMul::new(vec![
            tdscf_multiplier.clone(),
            tdscf_equation.clone(),
        ])?)?);
        lag_terms.push(Trace::new(MatrixMul::new(vec![
            idemp_multiplier.clone(),
            subtract_exprs(idempotency.clone(), density_matrix.clone())?,
        ])?)?);

        // The first term in Equation (98), J. Chem. Phys. 129, 214108 (2008)
        let set: HashSet<Arc<dyn Expr>> = [density_a].into_iter().collect();
        let generalized_energy_a = generalized_energy
            .differentiate(&perturbation_a)?
            .remove(&set)?;

        // The time-averaged quasi-energy derivative Lagrangian
        let lagrangian_dao = subtract_exprs(generalized_energy_a, Add::new(lag_terms)?)?;

        Ok(Self {
            perturbation_a: Some(perturbation_a),
            density_matrix,
            overlap_matrix,
            lambda,
            tdscf_multiplier,
            tdscf_equation,
            zeta,
            idemp_multiplier,
            idempotency,
            lagrangian_dao,
        })
    }

    //#[inline]
    //pub fn new_static(
    //    density_matrix: Arc<dyn Expr>,
    //    overlap_matrix: Option<Arc<dyn Expr>>,
    //    one_elec_operators: &[Arc<dyn Expr>],
    //    two_elec_operator: Option<Arc<dyn Expr>>,
    //    xc_energy: Option<Arc<dyn Expr>>,
    //    xc_potential: Option<Arc<dyn Expr>>,
    //    h_nuc: Option<Arc<dyn Expr>>,
    //) -> Result<Self, TinnedError> {
    //}

    #[inline]
    fn build_energy_and_fock(
        density_matrix: Arc<dyn Expr>,
        one_elec_operators: &[Arc<dyn Expr>],
        two_elec_operator: Option<Arc<dyn Expr>>,
        xc_energy: Option<Arc<dyn Expr>>,
        xc_potential: Option<Arc<dyn Expr>>,
        h_nuc: Option<Arc<dyn Expr>>,
    ) -> Result<(Arc<dyn Expr>, Arc<dyn Expr>), TinnedError> {
        if !is_expr_type::<WfnParameter>(&density_matrix) {
            return Err(expression_error(
                "Invalid type of density matrix",
                &density_matrix,
                None,
            ));
        }

        let mut energy_terms = Vec::with_capacity(one_elec_operators.len() + 3);
        let mut fock_terms = Vec::with_capacity(one_elec_operators.len() + 2);

        for op in one_elec_operators {
            fock_terms.push(op.clone());

            let matmul = MatrixMul::new(vec![op.clone(), density_matrix.clone()])?;
            let trace = Trace::new(matmul)?;
            energy_terms.push(trace);
        }

        if let Some(op) = &two_elec_operator {
            if let Some(two_elec_op) = downcast_from_arc::<TwoElecOperator>(op) {
                energy_terms.push(TwoElecEnergy::builder_from_operator(two_elec_op).build()?);
            } else {
                return Err(expression_error("Invalid two-electron operator", op, None));
            }
            fock_terms.push(op.clone());
        }

        if let Some(energy) = &xc_energy {
            energy_terms.push(energy.clone());
        }
        if let Some(op) = &xc_potential {
            fock_terms.push(op.clone());
        }
        if let Some(energy) = &h_nuc {
            energy_terms.push(energy.clone());
        }

        let generalized_energy = Add::new(energy_terms)?;
        let fock_matrix = MatrixAdd::new(fock_terms)?;

        Ok((generalized_energy, fock_matrix))
    }

    #[inline]
    pub fn particular_density_solution(
        &self,
        density_freq: Arc<dyn Expr>,
    ) -> Result<Arc<dyn Expr>, TinnedError> {
        if let Some(density_omega) = downcast_from_arc::<WfnParameter>(&density_freq) {
            let set: HashSet<Arc<dyn Expr>> = [density_freq].into_iter().collect();
            let diff_idempotency =
                differentiate_expr(&self.idempotency, density_omega.derivative())?.remove(&set)?;
            let anticomm_idemp_dm = if let Some(overlap) = &self.overlap_matrix {
                s_anticommutator(
                    diff_idempotency.clone(),
                    self.density_matrix.clone(),
                    overlap.clone(),
                )?
            } else {
                anticommutator(diff_idempotency.clone(), self.density_matrix.clone())?
            };

            subtract_exprs(anticomm_idemp_dm, diff_idempotency)
        } else {
            Err(expression_error(
                "Invalid type density matrix",
                &density_freq,
                None,
            ))
        }
    }

    #[inline]
    pub fn linear_response_rhs(
        &self,
        density_freq: Arc<dyn Expr>,
        density_part: Arc<dyn Expr>,
    ) -> Result<Arc<dyn Expr>, TinnedError> {
        if let Some(density_omega) = downcast_from_arc::<WfnParameter>(&density_freq) {
            let diff_tdscf = differentiate_expr(&self.tdscf_equation, density_omega.derivative())?;
            let map: HashMap<Arc<dyn Expr>, Arc<dyn Expr>> =
                std::iter::once((density_freq, density_part)).collect();

            Ok(diff_tdscf.replace(&map))
        } else {
            Err(expression_error(
                "Invalid type density matrix",
                &density_freq,
                None,
            ))
        }
    }
}

impl LagrangianInternal for LagrangianDao {
    #[inline]
    fn is_non_zero_sum_freqs(
        &self,
        exten_perturbations: &[Arc<Perturbation>],
        inten_perturbations: &[Arc<Perturbation>],
        num_tol: Option<NumberTolerance>,
    ) -> Result<bool, TinnedError> {
        let freq_sum_ext = sum_pert_frequencies(&exten_perturbations).map_err(|e| {
            generic_error(
                "Sum of extensive perturbations' frequencies failed",
                Some(Box::new(e)),
            )
        })?;
        let freq_sum_int = sum_pert_frequencies(&inten_perturbations).map_err(|e| {
            generic_error(
                "Sum of intensive perturbations' frequencies failed",
                Some(Box::new(e)),
            )
        })?;
        // Here we need to inlcude the frequency of `perturbation_a`, which can
        // either be extensive or intensive
        let mut terms = vec![freq_sum_ext, freq_sum_int];

        if let Some(pert) = &self.perturbation_a {
            terms.push(pert.frequency().clone());
        }

        let total_freq = Add::new(terms).map_err(|e| {
            generic_error(
                "Sum of all perturbations' frequencies failed",
                Some(Box::new(e)),
            )
        })?;

        Ok(!is_zero_expr(&total_freq, num_tol))
    }

    #[inline]
    fn eliminate_wfn_parameter(
        &self,
        lagrangian: &Arc<dyn Expr>,
        exten_perturbations: &[Arc<Perturbation>],
        min_wfn_order: u32,
    ) -> Result<Arc<dyn Expr>, TinnedError> {
        lagrangian.eliminate(&self.density_matrix, exten_perturbations, min_wfn_order)
    }

    #[inline]
    fn eliminate_lag_multipliers(
        &self,
        lagrangian: &Arc<dyn Expr>,
        exten_perturbations: &[Arc<Perturbation>],
        min_multiplier_order: u32,
    ) -> Result<Arc<dyn Expr>, TinnedError> {
        let mut result = lagrangian
            .eliminate(
                &self.tdscf_multiplier,
                exten_perturbations,
                min_multiplier_order,
            )
            .map_err(|e| {
                generic_error("Elimination of TDSCF multiplier failed", Some(Box::new(e)))
            })?;

        result = result
            .eliminate(
                &self.idemp_multiplier,
                exten_perturbations,
                min_multiplier_order,
            )
            .map_err(|e| {
                generic_error(
                    "Elimination of idempotency multiplier failed",
                    Some(Box::new(e)),
                )
            })?;

        // Replace "artificial" multipliers with real differentiated ones
        let replacements: HashMap<Arc<dyn Expr>, Arc<dyn Expr>> = HashMap::from([
            (self.tdscf_multiplier.clone(), self.lambda.clone()),
            (self.idemp_multiplier.clone(), self.zeta.clone()),
        ]);

        Ok(result.replace(&replacements))
    }
}

#[typetag::serde]
impl Lagrangian for LagrangianDao {
    #[inline]
    fn get_lagrangian(&self) -> &Arc<dyn Expr> {
        &self.lagrangian_dao
    }
}
